File: Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;

        auto pt1 = head, pt2 = head->next;
        pt1->next = nullptr;

        while (pt2) {
            auto pt3 = pt2->next;
            pt2->next = pt1;
            pt1 = pt2;
            pt2 = pt3;
        }

        return pt1;
    }
};

File: Reverse Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }

    pt1 := head
    pt2 := head.Next
    pt1.Next = nil

    for pt2 != nil {
        pt3 := pt2.Next
        pt2.Next = pt1
        pt1 = pt2
        pt2 = pt3
    }
    return pt1
}
